### PR TITLE
Fix processing of external add proposals 

### PR DESCRIPTION
**Pull request by stefanwire**
https://wearezeta.atlassian.net/browse/FS-921

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] I updated **changelog.d** subsections with one or more entries with the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):

### CODE REVIEW

**Review by smatting — APPROVED** [review]

**Review by pcapriotti — APPROVED** [review]
Looks good, but I've added some minor comments. Feel free to merge anyway if some of these things are too hard to address.

File: services/brig/src/Brig/API/Internal.hs [hunk 1]
```diff
@@ -184,6 +187,24 @@ getConvIdByKeyPackageRef = runMaybeT . mapMaybeT wrapClientE . Data.keyPackageRe
 postKeyPackageRef :: KeyPackageRef -> KeyPackageRef -> Handler r ()
 postKeyPackageRef ref = lift . wrapClient . Data.updateKeyPackageRef ref
 
+-- Used by galley to update key package refs and also validate
+upsertKeyPackage :: NewKeyPackage -> Handler r KeyPackageRef
+upsertKeyPackage nkp = do
+  kp :: RawMLS KeyPackage <- decodeMLS' (kpData . nkpKeyPackage $ nkp) & hush & noteH "upsertKeyPackage: Cannot decode KeyPackage"
```
[the remark this answers is not in the document]
Comment (stefanwire) [reply services/brig/src/Brig/API/Internal.hs]: done.

File: services/galley/test/integration/API/MLS/Util.hs [hunk 2]
```diff
@@ -256,7 +256,10 @@ setupParticipants tmp SetupOptions {..} ns = do
     createClientsForUR (RemoteUser _) _ = DontCreateClients
 
 withLastPrekeys :: Monad m => State.StateT [LastPrekey] m a -> m a
-withLastPrekeys m = State.evalStateT m someLastPrekeys
+withLastPrekeys m = withTheseLastPrekeys someLastPrekeys m
+
+withTheseLastPrekeys :: Monad m => [LastPrekey] -> State.StateT [LastPrekey] m a -> m a
+withTheseLastPrekeys pkeys m = State.evalStateT m pkeys
```
[the remark this answers is not in the document]
Comment (stefanwire) [reply services/galley/test/integration/API/MLS/Util.hs]: done.

[the remark this answers is not in the document]
Comment (stefanwire) [reply services/galley/test/integration/API/MLS/Util.hs]: Done.

File: services/galley/src/Galley/API/MLS/Message.hs [hunk 3]
```diff
@@ -511,25 +514,58 @@ applyProposalRef conv groupId epoch (Ref ref) = do
...
+    Nothing -> do
+      -- external add proposal for a new key package unknown to the backend
+      ci <-
+        either
+          (throw . mlsProtocolError . (<>) "Could not derive identity from key package: ")
+          pure
+          (kpIdentity (rmValue kp))
+      let qcid = cidQualifiedClient ci
+      lconvId <- qualifyLocal convId
+      void $ addKeyPackageMapping lconvId (fst <$> qcid) (snd (qUnqualified qcid)) (KeyPackageData (rmRaw kp))
```
[the remark this answers is not in the document]
Comment (stefanwire) [reply services/galley/src/Galley/API/MLS/Message.hs]: done.

**Pull request merged**